<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Vignette 1: Basic workflow • INClock</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.9/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Vignette 1: Basic workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">INClock</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_1_basic_workflow.html">Vignette 1: Basic workflow</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_2_downstream_analyses.html">Vignette 2: Downstream analyses</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Vignette 1: Basic workflow</h1>
                        <h4 data-toc-skip class="author">Yilin Yang</h4>
            
            <h4 data-toc-skip class="date">2024-09-11</h4>
      

      <div class="d-none name"><code>vignette_1_basic_workflow.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="vignette-info">Vignette Info<a class="anchor" aria-label="anchor" href="#vignette-info"></a>
</h2>
<p>This vignette depicts a basic workflow of INClock that generates
dispersion parameters for each cell type / age group. Examples of
downstream analyses can be found in Vignette 2.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://eddieyang1222.github.io/INClock/">INClock</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-the-test-data">Prepare the test data<a class="anchor" aria-label="anchor" href="#prepare-the-test-data"></a>
</h2>
<p>We will use single-cell RNA sequencing data from <a href="https://tabula-muris-senis.ds.czbiohub.org/" class="external-link">Tabula Muris
Senis</a> as our test data. Specifically, it concerns with the bone
marrow cells of mouse at different ages. As the original dataset is too
large for demonstration, we will only take a subset of its most relevant
cell types and age groups, which includes one younger (3 months) and one
older (30 months) age groups, as well as several stem cell types, such
as hematopoietic precursor cell (HSC). Ideally, if you want to reproduce
results similar to the ones in our paper, you should probably ignore the
subsetting part of data preparation below. The dataset is loaded with <a href="https://fmicompbio.github.io/TabulaMurisSenisData/" class="external-link">TabulaMurisSenisData</a>,
which is already integrated into our package for test purposes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load test data from Tabula Muris Senis</span></span>
<span><span class="va">tms_marrow</span> <span class="op">&lt;-</span> <span class="fu">TabulaMurisSenisData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TabulaMurisSenisData/man/TabulaMurisSenisDroplet.html" class="external-link">TabulaMurisSenisDroplet</a></span><span class="op">(</span>tissues <span class="op">=</span> <span class="st">"Marrow"</span><span class="op">)</span></span>
<span><span class="va">tms_marrow_counts</span> <span class="op">&lt;-</span> <span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tms_marrow_counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tms_marrow_counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">)</span></span>
<span><span class="va">tms_marrow_counts</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">tms_marrow_counts</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset age groups and cell types (ignore if to reproduce results from paper)</span></span>
<span><span class="va">subset_ages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">$</span><span class="va">age</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"3m"</span>, <span class="st">"30m"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">subset_cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">$</span><span class="va">cell_ontology_class</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"hematopoietic precursor cell"</span>, <span class="st">"megakaryocyte-erythroid progenitor cell"</span>,</span>
<span>  <span class="st">"precursor B cell"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tms_marrow_counts</span> <span class="op">&lt;-</span> <span class="va">tms_marrow_counts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">subset_ages</span>, <span class="va">subset_cell_types</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="compute-the-manifold">Compute the manifold<a class="anchor" aria-label="anchor" href="#compute-the-manifold"></a>
</h2>
<p>We can simply call <code><a href="../reference/compute_manifold.html">compute_manifold()</a></code> to compute the
manifold matrix for our test data. The function only requires one input
files:</p>
<ul>
<li>Gene expression count matrix (Gene by Cell)
(<code>counts</code>)</li>
</ul>
<p>The matrix has to be in either <code>matrix</code> or
<code>dgCMatrix</code> format, as required by dependent packages such as
<a href="https://mohuangx.github.io/SAVER/" class="external-link">SAVER</a> and <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>.</p>
<p>Before running, we need to specify the <code>method</code> to use,
either <code>'SAVER'</code> or <code>'neighbor'</code>. If
<code>'SAVER'</code> is selected, then you have the following
customizable options:</p>
<ul>
<li>Number of CPU cores to use. Default is 1. (<code>ncores</code>)</li>
</ul>
<p>If <code>'neighbor'</code> is selected, then you have the following
customizable options:</p>
<ul>
<li>Number of features for finding variable features. Default is 3000.
(<code>nfeatures</code>)</li>
<li>Number of dimensions for finding nearest neighbors. Default is 20.
(<code>dims</code>)</li>
<li>Number of neighbors to output. Default is 20.
(<code>neighbors</code>)</li>
</ul>
<p>The number of features and dimensions typically don’t affect the
quality of the manifold much. The number of neighbors used, however, can
somewhat affect. To facilitate an optimal selection of number of
neighbors, it is recommended to go through Vignette 3 (WIP). From our
previous tests, datasets of less than 100,000 cells typically have an
optimal number of neighbors around 20 to 30.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the manifold with SAVER</span></span>
<span><span class="va">tms_marrow_manifold_saver</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_manifold.html">compute_manifold</a></span><span class="op">(</span><span class="va">tms_marrow_counts</span>, method <span class="op">=</span> <span class="st">"SAVER"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the manifold with neighbor-based method</span></span>
<span><span class="va">tms_marrow_manifold_neighbor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_manifold.html">compute_manifold</a></span><span class="op">(</span><span class="va">tms_marrow_counts</span>, method <span class="op">=</span> <span class="st">"neighbor"</span><span class="op">)</span></span></code></pre></div>
<p>The output will be the manifold matrix of the same dimensions as the
original count matrix, but of course it is not sparse any more. Be
cautious that this step will use relatively large amount of memory.</p>
</div>
<div class="section level2">
<h2 id="estimate-dispersion-parameters">Estimate dispersion parameters<a class="anchor" aria-label="anchor" href="#estimate-dispersion-parameters"></a>
</h2>
<p>Now that we have the manifold computed, we can call
<code>estimate_dispersion</code> to estimate the dispersion parameters.
The function takes the following inputs:</p>
<ul>
<li>Gene expression count matrix (Gene by Cell)
(<code>counts</code>)</li>
<li>Manifold matrix (<code>manifold</code>)</li>
<li>A vector of cell types each cell in the dataset belongs to
(<code>cell.types</code>)</li>
<li>A vector of age groups each cell in the dataset belongs to
(<code>ages</code>)</li>
</ul>
<p>It is also necessary to choose a variance <code>model</code> for the
estimation, which can be <code>'cCV'</code> (constant Coefficient of
Variation), <code>'cFF'</code> (constant Fano Factor), or
<code>'cVar'</code> (constant Variance). We recommend using constant CV,
which is also the default model to be used.</p>
<p>Besides, there are a few more options to customize the process:</p>
<ul>
<li>Number of CPU cores to use. Default is 1. (<code>ncores</code>)</li>
<li>Cell size normalization factors. Default uses mean library size
normalization. (<code>size.factor</code>)</li>
<li>A vector of cell types labels that will be used in the estimation
process. (<code>cell.types.to.use</code>)</li>
<li>A vector of age group labels that will be used in the estimation
process. (<code>ages.to.use</code>)</li>
<li>Minimum number of cells in a cell type / age group. Groups with
fewer cells than this number will be skipped. Default is 10.
(<code>cell.types.cutoff</code>)</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create cell type and age labels (for demonstration)</span></span>
<span><span class="va">tms_marrow_cell_types</span> <span class="op">&lt;-</span> <span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">$</span><span class="va">cell_ontology_class</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">subset_ages</span>, <span class="va">subset_cell_types</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">tms_marrow_ages</span> <span class="op">&lt;-</span> <span class="va">tms_marrow</span><span class="op">$</span><span class="va">Marrow</span><span class="op">$</span><span class="va">age</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">subset_ages</span>, <span class="va">subset_cell_types</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create cell type and age labels (to reproduce results from paper)</span></span>
<span><span class="co"># tms_marrow_cell_types &lt;- tms_marrow$Marrow$cell_ontology_class</span></span>
<span><span class="co"># tms_marrow_ages &lt;- tms_marrow$Marrow$age</span></span>
<span></span>
<span><span class="co"># Estimate the dispersion parameters using constant CV</span></span>
<span><span class="va">tms_marrow_dispersions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimate_dispersion.html">estimate_dispersion</a></span><span class="op">(</span><span class="va">tms_marrow_counts</span>, <span class="va">tms_marrow_manifold_neighbor</span>,</span>
<span>  <span class="va">tms_marrow_cell_types</span>, <span class="va">tms_marrow_ages</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"cCV"</span>, ncores <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, we are able to generate a list object containing multiple
dataframes, each of which corresponds to a cell type. In each dataframe,
the first column will be genes, and the later ones will be the
dispersion parameter values for each age group.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yilin Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
